In [1]:
import requests
import pandas as pd
import json
import matplotlib.pyplot as plt
import numpy as np
from UN_Comtrade import download_trade_data
from UN_Comtrade import search_product_code, product_codes_with_parent
import os
from getData import getTradingPartners
from getData import getData
pd.set_option('display.max_rows', 300) # specifies number of rows to show
pd.options.display.float_format = '{:40,.4f}'.format # specifies default number format to 4 decimal places
%matplotlib inline

### Example

In [3]:
getData('Denmark','201401-201512',freq='m')

https://comtrade.un.org/api/get?ps=201401,201402,201403,201404,201405&freq=M&r=208&p=276,752,839,579,528&cc=TOTAL&rg=all&px=HS&type=C&fmt=json&max=50000&head=M
https://comtrade.un.org/api/get?ps=201406,201407,201408,201409,201410&freq=M&r=208&p=276,752,839,579,528&cc=TOTAL&rg=all&px=HS&type=C&fmt=json&max=50000&head=M
https://comtrade.un.org/api/get?ps=201411,201412,201501,201502,201503&freq=M&r=208&p=276,752,839,579,528&cc=TOTAL&rg=all&px=HS&type=C&fmt=json&max=50000&head=M
https://comtrade.un.org/api/get?ps=201504,201505,201506,201507,201508&freq=M&r=208&p=276,752,839,579,528&cc=TOTAL&rg=all&px=HS&type=C&fmt=json&max=50000&head=M
https://comtrade.un.org/api/get?ps=201509,201510,201511,201512&freq=M&r=208&p=276,752,839,579,528&cc=TOTAL&rg=all&px=HS&type=C&fmt=json&max=50000&head=M
240 records downloaded and saved as tempoaryFile.csv.


**Overall categories**

**01-05  Animal & Animal Products**

* 1: Animals; live
* 2: Meat and edible meat offal
* 3: Fish and crustaceans, molluscs and other aquatic invertebrates
* 4: Dairy produce; birds' eggs; natural honey; edible products of animal origin, not elsewhere specified or included
* 5: Animal originated products; not elsewhere specified or included

**06-15  Vegetable Products**

* 6: Trees and other plants, live; bulbs, roots and the like; cut flowers and ornamental foliage
* 7: Vegetables and certain roots and tubers; edible
* 8: Fruit and nuts, edible; peel of citrus fruit or melons
* 9: Coffee, tea, mate and spices
* 10: Cereals
* 11: Products of the milling industry; malt, starches, inulin, wheat gluten
* 12: Oil seeds and oleaginous fruits; miscellaneous grains, seeds and fruit, industrial or medicinal plants; straw and fodder
* 13: Lac; gums, resins and other vegetable saps and extracts
* 14: Vegetable plaiting materials; vegetable products not elsewhere specified or included
* 15: Animal or vegetable fats and oils and their cleavage products; prepared animal fats; animal or vegetable waxes

**16-24  Foodstuffs**

* 16: Meat, fish or crustaceans, molluscs or other aquatic invertebrates; preparations thereof
* 17: Sugars and sugar confectionery
* 18: Cocoa and cocoa preparations
* 19: Preparations of cereals, flour, starch or milk; pastrycooks' products
* 20: Preparations of vegetables, fruit, nuts or other parts of plants
* 21: Miscellaneous edible preparations
* 22: Beverages, spirits and vinegar
* 23: Food industries, residues and wastes thereof; prepared animal fodder
* 24: Tobacco and manufactured tobacco substitutes

**25-27  Mineral Products**

* 25: Salt; sulphur; earths, stone; plastering materials, lime and cement
* 26: Ores, slag and ash
* 27: Mineral fuels, mineral oils and products of their distillation; bituminous substances; mineral waxes

**28-38  Chemicals & Allied Industries**

* 28: Inorganic chemicals; organic and inorganic compounds of precious metals; of rare earth metals, of radio-active elements and of isotopes
* 29: Organic chemicals
* 30: Pharmaceutical products
* 31: Fertilizers
* 32: Tanning or dyeing extracts; tannins and their derivatives; dyes, pigments and other colouring matter; paints, varnishes; putty, other mastics; inks
* 33: Essential oils and resinoids; perfumery, cosmetic or toilet preparations
* 34: Soap, organic surface-active agents; washing, lubricating, polishing or scouring preparations; artificial or prepared waxes, candles and similar articles, modelling pastes, dental waxes and dental preparations with a basis of plaster
* 35: Albuminoidal substances; modified starches; glues; enzymes
* 36: Explosives; pyrotechnic products; matches; pyrophoric alloys; certain combustible preparations
* 37: Photographic or cinematographic goods
* 38: Chemical products n.e.s.

**39-40  Plastics / Rubbers**

* 39: Plastics and articles thereof
* 40: Rubber and articles thereof

**41-43  Raw Hides, Skins, Leather, & Furs**

* 41: Raw hides and skins (other than furskins) and leather
* 42: Articles of leather; saddlery and harness; travel goods, handbags and similar containers; articles of animal gut (other than silk-worm gut)
* 43: Furskins and artificial fur; manufactures thereof

**44-49  Wood & Wood Products**

* 44: Wood and articles of wood; wood charcoal
* 45: Cork and articles of cork
* 46: Manufactures of straw, esparto or other plaiting materials; basketware and wickerwork
* 47: Pulp of wood or other fibrous cellulosic material; waste and scrap of paper or paperboard
* 48: Paper and paperboard; articles of paper pulp, of paper or paperboard
* 49: Printed books, newspapers, pictures and other products of the printing industry; manuscripts, typescripts and plans

**50-63  Textiles**

* 50: Silk
* 51: Wool, fine or coarse animal hair; horsehair yarn and woven fabric
* 52: Cotton
* 53: Vegetable textile fibres; paper yarn and woven fabrics of paper yarn
* 54: Man-made filaments; strip and the like of man-made textile materials
* 55: Man-made staple fibres
* 56: Wadding, felt and nonwovens, special yarns; twine, cordage, ropes and cables and articles thereof
* 57: Carpets and other textile floor coverings
* 58: Fabrics; special woven fabrics, tufted textile fabrics, lace, tapestries, trimmings, embroidery
* 59: Textile fabrics; impregnated, coated, covered or laminated; textile articles of a kind suitable for industrial use
* 60: Fabrics; knitted or crocheted
* 61: Apparel and clothing accessories; knitted or crocheted
* 62: Apparel and clothing accessories; not knitted or crocheted
* 63: Textiles, made up articles; sets; worn clothing and worn textile articles; rags

**64-67  Footwear / Headgear**

* 64: Footwear; gaiters and the like; parts of such articles
* 65: Headgear and parts thereof
* 66: Umbrellas, sun umbrellas, walking-sticks, seat sticks, whips, riding crops; and parts thereof
* 67: Feathers and down, prepared; and articles made of feather or of down; artificial flowers; articles of human hair

**68-71  Stone / Glass**

* 68: Stone, plaster, cement, asbestos, mica or similar materials; articles thereof
* 69: Ceramic products
* 70: Glass and glassware
* 71: Natural, cultured pearls; precious, semi-precious stones; precious metals, metals clad with precious metal, and articles thereof; imitation jewellery; coin

**72-83  Metals**

* 72: Iron and steel
* 73: Iron or steel articles
* 74: Copper and articles thereof
* 75: Nickel and articles thereof
* 76: Aluminium and articles thereof
* 78: Lead and articles thereof
* 79: Zinc and articles thereof
* 80: Tin; articles thereof
* 81: Metals; n.e.s., cermets and articles thereof
* 82: Tools, implements, cutlery, spoons and forks, of base metal; parts thereof, of base metal
* 83: Metal; miscellaneous products of base metal

**84-85  Machinery / Electrical**

* 84: Nuclear reactors, boilers, machinery and mechanical appliances; parts thereof
* 85: Electrical machinery and equipment and parts thereof; sound recorders and reproducers; television image and sound recorders and reproducers, parts and accessories of such articles

**86-89  Transportation**

* 86: Railway, tramway locomotives, rolling-stock and parts thereof; railway or tramway track fixtures and fittings and parts thereof; mechanical (including electro-mechanical) traffic signalling equipment of all kinds
* 87: Vehicles; other than railway or tramway rolling stock, and parts and accessories thereof
* 88: Aircraft, spacecraft and parts thereof
* 89: Ships, boats and floating structures

**90-97  Miscellaneous**

* 90: Optical, photographic, cinematographic, measuring, checking, medical or surgical instruments and apparatus; parts and accessories
* 91: Clocks and watches and parts thereof
* 92: Musical instruments; parts and accessories of such articles
* 93: Arms and ammunition; parts and accessories thereof
* 94: Furniture; bedding, mattresses, mattress supports, cushions and similar stuffed furnishings; lamps and lighting fittings, n.e.s.; illuminated signs, illuminated name-plates and the like; prefabricated buildings
* 95: Toys, games and sports requisites; parts and accessories thereof
* 96: Miscellaneous manufactured articles
* 97: Works of art; collectors' pieces and antiques
* 99: Commodities not specified according to kind

In [2]:
# Two support functions are available to help locate the right category to extract
# 1. search_product_code
search_results = search_product_code('oil', case=False)

In [3]:
SR = list(search_results.items())

In [4]:
SR[0]

('0306',
 'Crustaceans; in shell or not, live, fresh, chilled, frozen, dried, salted or in brine; smoked, cooked or not before or during smoking; in shell, steamed or boiled, whether or not chilled, frozen, dried, salted or in brine; edible flours, meals, pellets')

In [5]:
# 2. product_codes_with_parent
product_codes_with_parent('01')

{'0101': 'Horses, asses, mules and hinnies; live',
 '0102': 'Bovine animals; live',
 '0103': 'Swine; live',
 '0104': 'Sheep and goats; live',
 '0105': 'Poultry; live, fowls of the species gallus domesticus, ducks, geese, turkeys and guinea fowls',
 '0106': 'Animals; live, n.e.c. in chapter 01'}

### Getting the tradingPartners of a country

In [6]:
def getTradingPartners(reporter,year='latest',indicatorType = 'all'):
    
    # Loading in the data if necessary
    if 'tradingPartners' not in locals():
        tradingPartners = pd.read_csv('TradingPartners.csv',index_col = 0)
    
    years = [int(column) for column in tradingPartners.columns if column.isnumeric()]
    
    # Checking the year provided
    
    if year == 'latest':
        year = max(years) 
    else:
        if isinstance(year,str):
            year = int(year)
        if year not in years:
            raise ValueError('The year is not included in the range of possible values')           
    
    # Making sure that the specified reporter and 'indicator type' is in the correct format (Starting with a capital letter, and the rest being
    # lowercase characters).
    # Reporter
    characters = [i.upper() if j == 0 else i.lower() for j,i in enumerate(reporter)]
    reporter = ''.join(characters)
    # Indicator Type
    characters = [i.upper() if j == 0 else i.lower() for j,i in enumerate(indicatorType)]
    indicatorType = ''.join(characters)
    
    # Checking that the specified reporter is part of the available reporters.
    if reporter not in tradingPartners.Reporter.unique():
        raise ValueError ('The specified reporter is not in the available set of reporters.')
    
    countryPartners = tradingPartners[tradingPartners.Reporter==reporter]
    countryPartners = countryPartners.reset_index(drop=True)
    
    # Creating a subset of interest
    subsettedFrame = pd.concat([countryPartners.iloc[:,0:5],countryPartners[str(year)]],axis=1)

    # Useful for later
    indicators = subsettedFrame.Indicator.unique()
    whereAreYouAmount = [i for i,element in enumerate(indicators) if 'share' not in element.lower()]
    whereAreYouShare = [i for i,element in enumerate(indicators) if 'share' in element.lower()]
    
    # Extracting the partners based on the indicator type specified.

    if (indicatorType =='all') or (indicatorType == 'All'):

        # We are only interested in the data with actual amounts of value listed, not share data.
        subsettedFrameActualAmounts = subsettedFrame[(subsettedFrame.Indicator == indicators[whereAreYouAmount[0]])|\
                                                     (subsettedFrame.Indicator == indicators[whereAreYouAmount[1]])]

        # Adding imports and exports together
        aggregatedData = {country:round(sum(subsettedFrameActualAmounts[subsettedFrameActualAmounts.Partner==country]\
                                            [str(year)]),3) for country in subsettedFrameActualAmounts.Partner.unique()}

        # Sorted highest to lowest
        sortedAggregatedData = sorted(aggregatedData.items(), key = lambda kv: kv[1],reverse = True)

        # Only keeping top five
        sortedDataTopFive = sortedAggregatedData[0:5]

    elif (indicatorType == 'Import') or (indicatorType == 'Export'):
        # Subsetting the data
        subsettedFrameActualAmountsEither = subsettedFrame[((subsettedFrame.Indicator == indicators[whereAreYouAmount[0]])|\
                                                            (subsettedFrame.Indicator == indicators[whereAreYouAmount[1]])) & \
                                                            (subsettedFrame['Indicator Type'] == indicatorType)]
        # Sorting the subsetted data
        sortedData = subsettedFrameActualAmountsEither.sort_values(by=[str(year)],ascending = False).round({str(year):3})
        sortedData = sortedData.reset_index(drop=True)
        # Getting top-five
        sortedDataTopFive = [(sortedData.loc[obs].Partner,(sortedData[str(year)].loc[obs]/100)) for obs in np.arange(5)]
    else: 
        raise ValueError ('The indicator type specified is not known.')
        
    # Getting the shares of the trading partners
    if (indicatorType == 'Import') or (indicatorType == 'Export'):
        subsettedFrameShares = subsettedFrame[((subsettedFrame.Indicator == indicators[whereAreYouShare[0]])|\
                                              (subsettedFrame.Indicator == indicators[whereAreYouShare[1]])) & \
                                             (subsettedFrame[str(year)]) & (subsettedFrame['Indicator Type'] == indicatorType)]
        # Sorted in share-frame
        sortedShares = subsettedFrameShares.sort_values(by=[str(year)],ascending = False).round({str(year):3})
        sortedShares = sortedShares.reset_index(drop=True)
        
        # Arranging top five properly
        sortedSharesTopFive = [(sortedShares.loc[obs].Partner,(sortedShares[str(year)].loc[obs]/100)) for obs in np.arange(5)]
    else:
        
        # Export
        exportAmount = [i for i,element in enumerate(indicators) if ('share' not in element.lower()) \
                        & ('export' in element.lower())]
        exportShare = [i for i,element in enumerate(indicators) if ('share' in element.lower()) \
                        & ('export' in element.lower())]
        
        eAmount = np.array([a for a in subsettedFrame[subsettedFrame.Indicator==indicators[exportAmount[0]]][str(year)]\
                          if a > 0])
        eShare = np.array([a for a in subsettedFrame[subsettedFrame.Indicator==indicators[exportShare[0]]][str(year)]\
                          if a > 0])
        
        totalExport = np.mean(eAmount / (eShare / 100))
        
        # Import
        
        importAmount = [i for i,element in enumerate(indicators) if ('share' not in element.lower()) \
                        & ('import' in element.lower())]
        importShare = [i for i,element in enumerate(indicators) if ('share' in element.lower()) \
                        & ('import' in element.lower())]
        
        iAmount = np.array([a for a in subsettedFrame[subsettedFrame.Indicator==indicators[importAmount[0]]][str(year)]\
                          if a > 0])
        iShare = np.array([a for a in subsettedFrame[subsettedFrame.Indicator==indicators[importShare[0]]][str(year)]\
                          if a > 0])
        
        totalImport = np.mean(iAmount / (iShare / 100))
        
        # Aggregated
        
        sortedSharesTopFive = [(country[0],(country[1]/(totalExport+totalImport))) for country in sortedDataTopFive]
        
    return sortedDataTopFive,sortedSharesTopFive

In [3]:
# Period: either YYYY or YYYYMM format, specify start/end period in one string, separated by a hyphen (Ex. YYYYMM-YYYYMM or 
# YYYY-YYYY).
# Year: YYYY format or 'latest'
# indType: 'all', 'Export' or 'Import'
# Freq: 'M' or 'Y'
# Reporter: Any country
# Products: 'TOTAL' or a valid HS code

def getData(reporter, period,products ='TOTAL',year='latest', indType='all',freq = 'Y'):
    ########################################## Getting cleaned data ##########################################
    
    # Getting top five trading partners
    topFiveAmount,topFiveShares = getTradingPartners(reporter=reporter,year=year,indicatorType=indType)
    
    # Pulling out the data received
    amounts = [partner[1] for partner in topFiveAmount]
    partners = [partner[0] for partner in topFiveAmount]
    
    # Extracting data from ComTrade API
    download_trade_data('tempoaryFile.csv', period=period, frequency=freq.upper(), reporter=reporter, 
                    partner=partners, product=products, tradeflow=indType)
    
    # Reading in the data, to structure it properly.
    cleanData = pd.read_csv('tempoaryFile.csv',index_col=0,parse_dates=['periodDesc'])
    
    # Keeping only the relevant columns
    cleanData = cleanData[['TradeValue','cmdCode','cmdDescE','periodDesc','ptTitle','rgDesc','rtTitle']]
    
    # Renaming
    cleanData.columns = ['trade_value','product','product_description','date','partner','indicator_type','reporter']
    
    # Converting the unit to 'millions' USD
    cleanData['trade_value'] = cleanData.trade_value/1000000
    
    # Structuring the data correctly
    cleanData = cleanData.sort_values(by=['partner','indicator_type','date','product'])
    cleanData = cleanData.reset_index(drop=True)
    
    # Saving the structured data
    cleanData.to_csv('StructuredData.csv')

Check what **'Special Categories'** implies:

From this source: https://unstats.un.org/unsd/tradekb/Knowledgebase/50042/Areas-not-elsewhere-specified

*The partner "Special Categories" (code 839) is used by a reporting country if it does not want the partner breakdown to be disclosed. The use of this partner code depends on the combination of reporting country, trade flow and specific commodity.*

There is also a category named *Unspecified*, which  is quite self-explainatory.